stdlib/public/core/TypeDiscovery.swift

```diff
@@ -0,0 +1,41 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_silgen_name("swift_enumerateConformingTypesFromImage")
+private func _swift_enumerateTypes(fromImageAt imageAddress: UnsafeRawPointer? = nil, conformingTo protocolType: Any.Type, _ body: (Any.Type, inout Bool) throws -> Void) rethrows
+
+/// Get an array of all types in a given image that conform to the specified
+/// Swift protocol.
+///
+/// - Parameters:
+///   - imageAddress: A platform-specific pointer to the image of interest. The
+///     image must have been loaded into the current process. For the binary of
+///     the calling function, you can pass `#dsohandle`. For all binaries, pass
+///     `nil` (the default.)
+///   - protocolType: The protocol of interest. This protocol does not need to
+///     be declared in the image referenced by `imageAddress`.
+///   - body: A closure to invoke once per conforming type.
+///
+/// - Throws: Whatever is thrown by `body`.
+///
+/// This function walks all known types in the given image and checks each for
+/// conformance to `protocol`. Any types that do conform are instantiated and
+/// added to the resulting array.
+///
+/// - Bug: Generic types with constrained conformance to `protocolType` are not
+///   supported.
+///
+/// - Bug: Objective-C protocol conformance lookups are not supported.
+public func _enumerateTypes(fromImageAt imageAddress: UnsafeRawPointer? = nil, conformingTo protocolType: Any.Type, _ body: (Any.Type, _ stop: inout Bool) throws -> Void) rethrows {
```

Comment:
We effectively need some way to say `<T: AnyProtocol>` like we can say `<T: AnyObject>`. Otherwise, when satisfying the generic requirement, `T.Type` ends up being interpreted as `T.Protocol` and `T.Protocol` is not a `T.Type`, so you can't invoke any static members on it.